src/chainstate/stacks/db/blocks.rs

```diff
@@ -1130,169 +1056,337 @@ impl StacksChainState {
         }
     }
 
-    /// Load up a preprocessed microblock (processed or not)
-    pub fn load_staging_microblock(
-        blocks_conn: &DBConn,
+    /// Load up a block's microblock public key hash, staging or not
+    fn load_block_pubkey_hash(
+        block_conn: &DBConn,
+        block_path: &String,
         consensus_hash: &ConsensusHash,
         block_hash: &BlockHeaderHash,
+    ) -> Result<Option<Hash160>, Error> {
+        let pubkey_hash = match StacksChainState::load_staging_block_pubkey_hash(
+            block_conn,
+            consensus_hash,
+            block_hash,
+        )? {
+            Some(pubkey_hash) => pubkey_hash,
+            None => {
+                // maybe it's already processed?
+                let header = match StacksChainState::load_block_header(
+                    block_path,
+                    consensus_hash,
+                    block_hash,
+                )? {
+                    Some(block_header) => block_header,
+                    None => {
+                        // parent isn't available
+                        return Ok(None);
+                    }
+                };
+                header.microblock_pubkey_hash.clone()
+            }
+        };
+        Ok(Some(pubkey_hash))
+    }
+
+    /// Load up a preprocessed microblock's staging info (processed or not), but via
+    /// its parent anchored block's index block hash.
+    /// Don't load the microblock itself.
+    /// Ignores orphaned microblocks.
+    pub fn load_staging_microblock_info(
+        blocks_conn: &DBConn,
+        parent_index_block_hash: &StacksBlockId,
         microblock_hash: &BlockHeaderHash,
     ) -> Result<Option<StagingMicroblock>, Error> {
-        let sql = "SELECT * FROM staging_microblocks WHERE consensus_hash = ?1 AND anchored_block_hash = ?2 AND microblock_hash = ?3 AND orphaned = 0".to_string();
-        let args: &[&dyn ToSql] = &[&consensus_hash, &block_hash, &microblock_hash];
+        let sql = "SELECT * FROM staging_microblocks WHERE index_block_hash = ?1 AND microblock_hash = ?2 AND orphaned = 0 LIMIT 1".to_string();
+        let args: &[&dyn ToSql] = &[&parent_index_block_hash, &microblock_hash];
         let mut rows =
             query_rows::<StagingMicroblock, _>(blocks_conn, &sql, args).map_err(Error::DBError)?;
-        let len = rows.len();
-        match len {
-            0 => Ok(None),
-            1 => {
+
+        Ok(rows.pop())
```

Comment:
Same here -- the `query_row` function has the same semantics as just calling `pop()` on the SQL query result, but it avoids the vec.

Reply:
Addressed in ee12847